api/python/t4/formats.py

```diff
@@ -134,15 +134,36 @@ def search(cls, obj_type=None, meta=None, ext=None):
             ext: The filename extension for the data
                 If given, then if other methods fail or are not specified,
                 the handler(s) for the extension `ext` will be returned.
+                If other methods do not fail, any handlers that support the
+                specified extension will be moved to the front of the result
+                list.
 
         Returns:
             list: Formats in order of preference (latest-added first)
+
+        Raises:
+            QuiltException: if no matching formats are found
         """
-        # we want to retain order.
+        # Reasons to use lists and not sets:
+        # * we want to retain order, so recently added formats take precedence
+        # * at this scale, lists are faster than sets
         meta_fmts = cls.for_meta(meta)
         typ_fmts = cls.for_type(obj_type)
+        ext_fmts = cls.for_ext(ext)
         fmt_name = cls._get_name_from_meta(meta)
 
+        def bump_ext_fmts(fmts):
+            """return a new list from fmts, shifting those in ext_fmts to the start"""
```

Comment:
From a control flow/ readability point of view, can you try refactoring this and see if it improves. Something where you apply list intersection successively, i.e.:

```
ext_fmts = cls.for_ext(ext)
results = ext_fmts
if obj_type:
    ...
    results = [fmt for fmt in results if fmt in typ_fmts]
    ...
if fmt_name:
    ...
    results = [fmt for fmt in results if fmt in meta_fmts]
    ...

```

That will remove the nested `if` inside the `obj_type` check, reduce code duplication there, and removes the necessity for the `bump_ext_fmts` inner method.

Reply:
That would exclude `typ_fmts` and `meta_fmts` that don't match an extension.  We don't want to do that, because we need to be able to handle situations where:

* something looks like an extension, but is just '.something' at the end of the file -- like "available housing 2019-04-17 19:35:59.287858"
* something *is* an extension, but the user has specified a different format.

I can clean that up a bit, though, in any case.

Reply:
Done.